### ISSUE TITLE
CSS is stripped, but stylesheet links are preserved

### ISSUE THREAD

**Issue by aehlke** [mozilla/readability#380]
Looks like the intention is to remove all CSS from the document, per https://github.com/mozilla/readability/blame/a63e63c91ee060873e3db06babd217323ad3979b/Readability.js#L382 and stripping inline style attributes elsewhere. However stylesheets of the form `<link rel="stylesheet" property="stylesheet" href="foobar.css">` are preserved, which can lead to surprising results.

**Comment by andreskrey** [comment]
Could not reproduce. Can you provide an example HTML that fails to be parsed correctly?

I used the following markup:

```
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8"/>
    <link rel="stylesheet" property="stylesheet" href="foobar.css">

    <title>Remove trailing brs test</title>
</head>
<body>
<article>
    <h1>Lorem</h1>
    <link rel="stylesheet" property="stylesheet" href="foobar.css">

    <div>
        <br/>
        <p>Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
            tempor incididunt ut labore et dolore magna aliqua.</p>
        <br/><br/><p>Ut enim ad minim veniam,
        quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
        consequat.</p><br/>
        <p>Duis aute irure dolor in reprehenderit in voluptate velit esse
            cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
            proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>
    </div>
    <h2>Foo</h2>
    <div>
        <p>Tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
            quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
            consequat.</p>
        <p>Duis aute irure dolor in reprehenderit in voluptate velit esse
            cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
            proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>
    </div>
</article>
</body>
</html>
```

**Comment by aehlke** [comment]
I don't have one on hand right now, but are you certain the result for your example lacks both link tags? Perhaps the link tag needs to appear within a block that readability considers to be part of the article. I checked the latest code in master and there's still no explicit removal of link tags, or mention of the `"stylesheet"` attribute, so the only way those tags are getting removed is if they happen to fall outside the areas Readability is extracting.

**Comment by andreskrey** [comment]
Only if I move the `link` tag inside the `div` the node doesn't get removed. I'm not even sure if those tags are allowed outside the `head`.

I you provide a real world example, maybe we can come up with a patch and a test case.

**Comment by aehlke** [comment]
Just to note:

>A <link> tag can occur either in the head element or in the body element (or both), depending on whether it has a link type that is body-ok. For example, the stylesheet link type is body-ok, and therefore a <link rel="stylesheet"> is permitted in the body.
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link

**Comment by andreskrey** [comment]
I stand corrected :F

**Comment by zoehneto** [comment]
I just created a minimal sample demonstrating the issue:
```
<div></div>

<html>
<head>
<link href="main.css" rel="stylesheet" type="text/css">
</head>

<body>
    <div>
        <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla sit amet iaculis orci. Mauris nec neque vulputate, blandit libero non, ultrices ex. Ut consectetur, velit sit amet egestas feugiat, lorem purus ultricies erat, at mollis lacus neque dapibus augue. Fusce mi tellus, maximus ut aliquam id, euismod nec urna. Etiam volutpat aliquam lorem eget rhoncus. Aenean congue magna tortor, vitae volutpat magna mattis quis. Mauris eu massa non velit dapibus porttitor aliquet id erat. Suspendisse bibendum, felis id posuere lobortis, lacus felis mattis leo, vitae pulvinar nunc ipsum at diam. Etiam vel tempor justo. Integer eget nulla maximus, volutpat urna eget, tristique erat. Proin euismod leo ac cursus sollicitudin.
        </p>
    </div>
</body>
</html>
```
Now this is not valid HTML because of the `div` before the head element, but it is quite common to inject some advertising script or similar before the main content of the website. Also interesting is that replacing the second `div` (the one surrounding the paragraph) with an `article` tag also resolves this problem.

**Comment by zoehneto** [comment]
From what I can tell there are two reasons for the behaviour shown in this example:

1.  The `div` before the `html` leads to chrome interpreting the html as follows (the `link` element is moved to the `body`, which is allowed in [HTML 5](https://www.w3.org/TR/html5/document-metadata.html#the-link-element):
```
<html>
<head>
<style></style>
</head>
<body>
    <div></div>

    <link href="main.css" rel="stylesheet" type="text/css">

    <div>
        <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla sit amet iaculis orci. Mauris nec neque vulputate, blandit libero non, ultrices ex. Ut consectetur, velit sit amet egestas feugiat, lorem purus ultricies erat, at mollis lacus neque dapibus augue. Fusce mi tellus, maximus ut aliquam id, euismod nec urna. Etiam volutpat aliquam lorem eget rhoncus. Aenean congue magna tortor, vitae volutpat magna mattis quis. Mauris eu massa non velit dapibus porttitor aliquet id erat. Suspendisse bibendum, felis id posuere lobortis, lacus felis mattis leo, vitae pulvinar nunc ipsum at diam. Etiam vel tempor justo. Integer eget nulla maximus, volutpat urna eget, tristique erat. Proin euismod leo ac cursus sollicitudin.
        </p>
    </div>

</body>
</html>
```

2. If the `div` is replaced by an `article` this issue doesn't exist, since the content of the `article` has a higher score  and is therefore selected (correctly ignoring the `link` element above it).

So in conclusion it seems the issue is that readability does nothing about `link` elements in the part of the `body` which is selected. This is often not a problem, since stylesheet references in the body are rare outside of malformed HTML, but should still be fixed.

**Comment by aehlke** [comment]
Thank you 😄
